TokenSet.cs

```diff
@@ -0,0 +1,578 @@
+ * B.Kowarsch and R.Sutcliffe "Modula-2 Revision 2010" (M2R10).
+ *
+ * In translator mode, M2Sharp translates Modula-2 source to C# source files.
+ * In compiler mode, M2Sharp compiles Modula-2 source via C# source files
+ * to object code or executables using the host system's C# compiler.
+ *
+ * @repository
+ *
+ * https://github.com/m2sf/m2sharp
+ *
+ * @file
+ *
+ * TokenSet.cs
+ * 
+ * A set of Token type variables which is used to pass First and Follow sets
+ *
+ * @license
+ *
+ * M2Sharp is free software: you can redistribute and/or modify it under the
+ * terms of the GNU Lesser General Public License (LGPL) either version 2.1
+ * or at your choice version 3 as published by the Free Software Foundation.
+ * However, you may not alter the copyright, author and license information.
+ *
+ * M2Sharp is distributed in the hope that it will be useful,  but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  Read the license for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with M2Sharp. If not, see <https://www.gnu.org/copyleft/lesser.html>.
+ *
+ * NB: Components in the domain part of email addresses are in reverse order.
+ */
+
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace org.m2sf.m2sharp
+{
+
+    public class TokenSet : ITokenSet {
+
+        /* Lexeme Table */
+        public static string[] lexemeTable = {
+  /* Null Token */
+  
+  "UNKNOWN",
+  
+  /* Reserved Words */
+
+  "AND", 
+  "ARGLIST",
+  "ARRAY", 
+  "BEGIN", 
+  "BY",
+  "CASE",
+  "CONST",
+  "DEFINITION",
+  "DIV",
+  "DO",
+  "ELSE",
+  "ELSIF",
+  "END",
+  "EXIT",
+  "EXPORT",
+  "FOR",
+  "FROM",
+  "IF",
+  "IMPLEMENTATION",
+  "IMPORT",
+  "IN",
+  "LOOP",
+  "MOD",
+  "MODULE",
+  "NOT",
+  "OF",
+  "OR",
+  "POINTER",
+  "PROCEDURE",
+  "QUALIFIED",
+  "RECORD",
+  "REPEAT",
+  "RETURN",
+  "SET",
+  "THEN",
+  "TO",
+  "TYPE",
+  "UNTIL",
+  "VAR",
+  "WHILE",
+  "WITH",
+  
+  /* Identifiers */
+  
+  "IDENTIFIER",
+  
+  /* Literals */
+  
+  "STRING-LITERAL",
+  "INTEGER-LITERAL",
+  "REAL-LITERAL",
+  "CHAR-LITERAL",
+  
+  "MALFORMED-STRING",
+  "MALFORMED-INTEGER",
+  "MALFORMED-REAL",
+  
+  /* Pragmas */
+  
+  "PRAGMA",
+  
+  /* Special Symbols */
+  
+  "PLUS",
+  "MINUS",
+  "EQUAL",
+  "NOTEQUAL",
+  "LESS-THAN",
+  "LESS-OR-EQUAL",
+  "GREATER-THAN",
+  "GREATER-OR-EQUAL",
+  "ASTERISK",
+  "SOLIDUS",
+  "BACKSLASH",
+  "ASSIGNMENT",
+  "COMMA",
+  "PERIOD",
+  "COLON",
+  "SEMICOLON",
+  "RANGE",
+  "DEREF",
+  "VERTICAL-BAR",
+  "LEFT-PAREN",
+  "RIGHT-PAREN",
+  "LEFT-BRACKET",
+  "RIGHT-BRACKET",
+  "LEFT-BRACE",
+  "RIGHT-BRACE",
+  "END-OF-FILE",
+  
+  /* out-of-range guard */
+  
+  "\0"
+
+}; /* end m2c_token_name_table */
+
+        public static int segmentCount = (Enum.GetNames(typeof(Token)).Length / 32) + 1;
+
+        public struct opaque

```

Comment:
We should use a more descriptive name here, possibly also change the struct to something else.

An opaque type is a type whose name is defined in a public interface and whose implementation is hidden. Instances of the type are opaque pointers to hidden data, that is, the pointer is exposed to clients but the structure that it points to is unknown to them.

The C syntax to declare opaque types is not self explanatory. It isn't immediately obvious what the type declaration means. For this reason we use a naming convention in C that makes it explicit in the name that the type is an opaque type. Thus a type we might otherwise have called `tokenset_t` is then called `tokenset_opaque_t` instead.

Both the structure and the naming needs to be appropriately adjusted in the C# version. In C# the primary means to hide data is the class. The struct is simply a completely hidden type, it is not exposed to the client. The instance of the hidden type is an instance variable.

The advantage of keeping the struct is that it can be mapped to the initialisation data which we want to be struct based so we can initialise it using a structured literal like { 0x0, 0x0, 0x3, 2 }. Otherwise we could simply declare separate instance vars, one for each segment, one for the count.

```csharp
private struct TokenSetBits {
  uint bits64to95;
  uint bits32to63;
  uint bits0to31;
  uint count;
} /* end TokenSetBits */
```

alternatively

```csharp
private struct TokenSetBits {
  uint[SegmentCount] segments;
  uint count;
} /* end TokenSetBits */
```
which allows us to address the segments by index (desirable), but then SegmentCount has to be a compile time constant and the value needs to be calculated at compile time (absolute must), not at runtime.